docker/android/android/make.py

```diff
@@ -0,0 +1,513 @@
+'''
+    make
+    ====
+
+    utilities to process makefiles. this parser is not sophisticated
+    nor correct, but it tries to avoid a few common pitfalls by
+    handling conditional blocks, and first separating all conditional
+    blocks into sections, and then parsing comment blocks within those
+    sections.
+
+    validate conditional directives are:
+    - ifeq
+    - ifneq
+    - ifdef
+    - ifndef
+    - else
+    - endif
+
+    makefiles are whitespace-sensitive, but not with leading whitespace
+    for conditional directives. for example, this is valid (replacing the
+    spaces with tabs):
+
+        # ---------------
+        # Section 1.
+        # ---------------
+        ifneq ($(USE_A),)
+            # -----------
+            # Section 2.
+            # -----------
+            ifneq ($(USE_B),)
+                SOURCES=b.cc
+            else
+                SOURCES=a.cc
+            endif
+        else
+            SOURCES=c.cc
+        endif
+
+    our goals are fairly different from a regular parser: we want to detect
+    and excise sections based on the comments, while ensuring that we do
+    not produce invalid output. other than unbalanced conditional directives,
+    we do not actually care about the actual contents.
+
+    for this, we use a 3 step parsing approach:
+    1. break up document into blocks separated by directives
+        - each block can be a regular or directive block
+        - directive blocks have a start and end directive as well as contents
+        - directives can be infinitely nested: the contents can also be a list
+    2. break each text block based on comment sections
+    3. group blocks within comment sections
+
+    for example, in the above, we want the entire makefile to be inside the
+    section 1 comment block, so removing it would remove that whole tree.
+    similarly, the inner directive block should be inside the section 2
+    comment block. we would therefore produce something like this:
+
+        CommentBlock: Section 1
+          Directive Block:
+            start=ifneq ($(USE_A),)
+            end=endif
+            children:
+              CommentBlock: Section 2
+                Directive Block:
+                  start=ifneq ($(USE_B),)
+                  end=endif
+                  children:
+                    Block: `SOURCES=b.cc\nelse\nSOURCES=a.cc`
+              Block: `else\nSOURCES=c.cc`
+'''
+
+import re2 as re
```

Comment:
We desperately need `re2` since some specific edge-cases work abysmally with backtracking regular expressions. This can mean it will take **hours** to match a header with Python's built-in `re` library for what takes milliseconds or less with `re2`.